### PR TITLE
Fix warning for unused variables in or pattern (issue #67691)

### PR DESCRIPTION
**Pull request by sapir**
Is this a good way to fix it?

Also, the tests fail, the "fixed" code output says `{ i, j }` instead of `{ i, j: _ }`, how can I fix that?

### CODE REVIEW

File: src/test/ui/lint/issue-67691-unused-field-in-or-pattern.rs [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+pub enum MyEnum {
+    A { i: i32, j: i32 },
+    B { i: i32, j: i32 },
+}
+
+pub fn foo(x: MyEnum) {
+    use MyEnum::*;
+
+    match x {
+        A { i, j } | B { i, j } => { //~ ERROR unused variable
```
Comment (Centril) [inline src/test/ui/lint/issue-67691-unused-field-in-or-pattern.rs]: I'd like to see an additional case `Some(A { i, j } | B { i, j })` to make sure this works for `#![feature(or_patterns)]` as well. (Applies to the next test also.)

Comment (sapir) [reply src/test/ui/lint/issue-67691-unused-field-in-or-pattern.rs]: Along with the warnings, I get an "error: or-patterns are not fully implemented yet", so should I include that in the expected stderr?

**Comment by sapir** [comment]
Why aren't the `ref` keywords being included in the suggestion? The test for #54180 is passing, and afaict the span being passed to multipart_suggestion does include the `ref` keyword.

File: src/test/ui/lint/issue-67691-unused-field-in-or-pattern.rs (hunk 1 above)
Comment (Centril) [reply src/test/ui/lint/issue-67691-unused-field-in-or-pattern.rs]: I think we can wait with landing this PR until https://github.com/rust-lang/rust/pull/67668 lands and then there shouldn't be such errors.

**Comment by Centril** [comment]
Nested or-patterns now work on nightly, so we should be able to add tests for those now as well, so this PR is now unblocked.

File: src/librustc_passes/liveness.rs [hunk 2]
```diff
@@ -1512,35 +1517,48 @@ impl<'tcx> Liveness<'_, 'tcx> {
...
                 let mut err = self.ir.tcx.struct_span_lint_hir(
                     lint::builtin::UNUSED_VARIABLES,
-                    hir_id,
-                    spans.clone(),
+                    first_hir_id,
+                    hir_ids_and_spans.iter().map(|(_, sp)| *sp).collect::<Vec<_>>(),
                     &format!("unused variable: `{}`", name),
                 );
 
                 if self.ir.variable_is_shorthand(var) {
```
Comment (matthewjasper) [inline src/librustc_passes/liveness.rs]: This should be checked for each variable in the pattern: `A{ x } | B { f: x }` is allowed.

**Comment by Centril** [comment]
@sapir Any updates? // Triage

**Comment by sapir** [comment]
I'll try again

File: src/librustc_passes/liveness.rs (hunk 2 above)
Comment (sapir) [reply src/librustc_passes/liveness.rs]: Which warning should be printed in that case?

Comment (sapir) [reply src/librustc_passes/liveness.rs]: "try ignoring these fields..." "...and prefixing these with an underscore"?

File: src/test/ui/lint/issue-67691-unused-field-in-or-pattern.stderr [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+error: unused variable: `i`
+  --> $DIR/issue-67691-unused-field-in-or-pattern.rs:70:24
+   |
+LL |         MixedEnum::A { ref i } | MixedEnum::B(ref i) => {
+   |                        ^^^^^                  ^^^^^
+   |
+help: try ignoring the field
+   |
+LL |         MixedEnum::A { i: _ } | MixedEnum::B(_i) => {
```
Comment (matthewjasper) [inline src/test/ui/lint/issue-67691-unused-field-in-or-pattern.stderr]: This should be suggesting `MixedEnum::A { i: _ } | MixedEnum::B(_)` or `MixedEnum::A { i: ref _i } | MixedEnum::B(ref _i)`

Comment (sapir) [reply src/test/ui/lint/issue-67691-unused-field-in-or-pattern.stderr]: ok

**Comment by sapir** [comment]
I think the rustfix test is failing because rustfix doesn't apply multipart suggestions: https://github.com/rust-lang/rustfix/commit/42fd66debb29125356d42aad0eb88f72ce99bc0e

**Comment by joelpalmer** [comment]
Ping from Triage: Any updates @sapir?

**Comment by sapir** [comment]
@joelpalmer I don't know what I should do about the multipart suggestions?

**Comment by matthewjasper** [comment]
`multipart_suggestion` is correct. You;ll have to remove `run-rustfix` from the test and leave a FIXME.

**Comment by joelpalmer** [comment]
Ping from Triage: @sapir Any updates?

**Comment by sapir** [comment]
@joelpalmer I did as @matthewjasper suggested regarding multipart_suggestions, so the tests pass now. Also I rebased to current master and fixed some conflicts.

**Pull request merged**